pkg/chains/formats/slsa/v2/taskrun/taskrun.go

```diff
@@ -0,0 +1,93 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package taskrun
+
+import (
+	"context"
+	"reflect"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	"github.com/tektoncd/chains/pkg/chains/formats/slsa/extract"
+	slsav1 "github.com/tektoncd/chains/pkg/chains/formats/slsa/v1/taskrun"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/logging"
+)
+
+// BuildConfig is the custom Chains format to fill out the
+// "buildConfig" section of the slsa-provenance predicate
+type BuildConfig struct {
+	TaskSpec *v1beta1.TaskSpec `json:"taskSpec"`
+}
+
+func GenerateAttestation(builderID string, tro *objects.TaskRunObject, ctx context.Context) (interface{}, error) {
+	logger := logging.FromContext(ctx)
+	subjects := extract.SubjectDigests(tro, logger)
+
+	mat, err := slsav1.Materials(tro, logger)
+	if err != nil {
+		return nil, err
+	}
+	att := intoto.ProvenanceStatement{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: slsa.PredicateSLSAProvenance,
+			Subject:       subjects,
+		},
+		Predicate: slsa.ProvenancePredicate{
+			Builder: slsa.ProvenanceBuilder{
+				ID: builderID,
+			},
+			BuildType:   tro.GetGVK(),
+			Invocation:  invocation(tro),
+			BuildConfig: BuildConfig{TaskSpec: tro.Status.TaskSpec},
+			Metadata:    slsav1.Metadata(tro),
+			Materials:   mat,
+		},
+	}
+	return att, nil
+}
+
+// invocation describes the event that kicked off the build
+// we currently don't set ConfigSource because we don't know
+// which material the Task definition came from
+func invocation(tro *objects.TaskRunObject) slsa.ProvenanceInvocation {
+	i := slsa.ProvenanceInvocation{}
+	if p := tro.Status.Provenance; p != nil {
+		i.ConfigSource = slsa.ConfigSource{
+			URI:        p.ConfigSource.URI,
+			Digest:     p.ConfigSource.Digest,
+			EntryPoint: p.ConfigSource.EntryPoint,
+		}
+	}
+	i.Parameters = invocationParams(tro)
+	return i
+}
+
+// invocationParams adds all fields from the task run object except
+// TaskRef or TaskSpec since they are in the ConfigSource or buildConfig.
+func invocationParams(tro *objects.TaskRunObject) map[string]any {
+	var iParams map[string]any = make(map[string]any)
+	skipFields := sets.NewString("TaskRef", "TaskSpec")
+	v := reflect.ValueOf(tro.Spec)
+	for i := 0; i < v.NumField(); i++ {
+		fieldName := v.Type().Field(i).Name
+		if !skipFields.Has(v.Type().Field(i).Name) {
+			iParams[fieldName] = v.Field(i).Interface()
+		}
+	}
+	return iParams
+}
```

Comment:
Here, I am choosing all fields from the TaskRunSpec struct except `TaskRef` and `TaskSpec`. Please let me know if this is ok?

Alternatively, I could append specific fields but that would require that we update this part every time Tekton pipelines makes an update to the `TaskRunSpec`. Given that is not too frequent, it might be ok to take that strategy as well.